src/packet_sources/macos.rs

```diff
@@ -0,0 +1,274 @@
+            process_name: Some(packet.process_name),
+        });
+    } else {
+        return Err(prost::DecodeError::new("Failed to decode packet"));
+    }
+}
+
+pub fn copy_dir(src: &Path, dst: &Path) -> Result<()> {
+    for entry in src.read_dir()? {
+        let entry = entry?;
+        let ty = entry.file_type()?;
+        let src_path = entry.path();
+        let dst_path = dst.join(entry.file_name());
+        if ty.is_dir() {
+            fs::create_dir_all(&dst_path)?;
+            copy_dir(&src_path, &dst_path)?;
+        } else {
+            fs::copy(&src_path, &dst_path)?;
+        }
+    }
+    Ok(())
+}
+
+pub struct PipeServer {
+    tx: pipe::Sender,
+    rx: pipe::Receiver,
+    path: PathBuf,
+}
+
+impl PipeServer {
+    pub fn new(fifo_name: &str) -> Result<Self> {
+        let home_dir = home_dir().unwrap();
+        let fifo_path = Path::new(&home_dir).join(format!("Downloads/{}.pipe", &fifo_name));
+        match mkfifo(&fifo_path, Mode::S_IRWXU) {
+            Ok(_) => println!("created {:?}", fifo_path),
+            Err(err) => println!("Error creating fifo: {}", err),
+        }
+        Ok(PipeServer {
+            tx: pipe::OpenOptions::new().open_sender(&fifo_path)?,
+            rx: pipe::OpenOptions::new().open_receiver(&fifo_path)?,
+            path: fifo_path,
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum MacosIpcRecv {
+    Packet {
+        data: Vec<u8>,
+        process_name: Option<String>,
+    },
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum MacosIpcSend {
+    Packet(Vec<u8>),
+    SetIntercept(InterceptConf),
+}
+
+pub struct MacosConf;
+
+#[async_trait]
+impl PacketSourceConf for MacosConf {
+    type Task = MacosTask;
+    type Data = UnboundedSender<MacosIpcSend>;
+
+    fn name(&self) -> &'static str {
+        "Macos proxy"
+    }
+
+    async fn build(
+        self,
+        net_tx: Sender<NetworkEvent>,
+        net_rx: Receiver<NetworkCommand>,
+        sd_watcher: broadcast::Receiver<()>,
+    ) -> Result<(MacosTask, Self::Data)> {
+        // #[cfg(target_os = "macos")]
+        // tokio::spawn(async {
+        //     use tokio::signal;
+        //     if let Ok(_) = signal::ctrl_c().await {
+        //         let _ = Command::new("networksetup")
+        //             .args(["-setdnsservers", "Wi-Fi", "empty"])
+        //             .output();
+        //
+        //         let _ = Command::new("route")
+        //             .args(["-n", "delete", "default"])
+        //             .output();
+        //         let _ = Command::new("route")
+        //             .args(["-n", "add", "default", "192.168.1.1"])
+        //             .output();
+        //         process::exit(0);
+        //     }
+        // });
+
+        let executable_path = "/Applications/MitmproxyAppleTunnel.app/";
+        copy_dir(
+            Path::new("../apple-tunnel/MitmproxyAppleTunnel.app/"),
+            Path::new(executable_path),
+        )?;
```

Comment:
Here's what we do on Windows:

1. We first build `windows-redirector.exe` [here](https://github.com/mitmproxy/mitmproxy_rs/blob/d7df8aa777d7ce149476a0f7b4648665d3fb8a38/.github/workflows/ci.yml#L109-L111).
2. The we build the rest. `ffi/build.rs` copies over the executable + WinDivert into the Python package in [build.rs](https://github.com/mitmproxy/mitmproxy_rs/blob/d7df8aa777d7ce149476a0f7b4648665d3fb8a38/ffi/build.rs). This ensures they are available in the Python wheel.
3. We check for existence of the executable in `start_os_proxy` (before any of the async tasks are spun up.
4. Finally, we blindly execute the app in [packet_sources/windows.rs](https://github.com/mitmproxy/mitmproxy_rs/tree/main/src/packet_sources)


I think we should do the same for macOS. Does that make sense? :)

Reply:
yes absolutely! The only problem is that the Tunnel in this case is not written in rust so it has to be compiled previously by xcode, inside build.rs I then just make the copy of the files